1.2-Introduction-to-Musig.ipynb

```diff
@@ -153,31 +154,56 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "# Generate nonces\n",
-    "k0 =  # TODO: implement\n",
-    "k1 =  # TODO: implement\n",
-    "k2 =  # TODO: implement\n",
-    "R0 =  # TODO: implement\n",
+    "# Generate nonces and nonce points\n",
+    "# Method: generate_schnorr_nonce() with no argument generates a random nonce, or it can be provided\n",
+    "# a seed to generate a nonce deterministically.\n",
+    "k1 = generate_schnorr_nonce(b'seed1')\n",
+    "k2 = generate_schnorr_nonce(b'seed2')\n",
+    "k3 = generate_schnorr_nonce(b'seed3')\n",
+    "# Method: use get_pubkey() to get the associated nonce point.\n",
     "R1 =  # TODO: implement\n",
     "R2 =  # TODO: implement\n",
+    "R3 =  # TODO: implement\n",
     "\n",
-    "# Nonce point commitment round\n",
-    "R0_digest = hashlib.sha256(R0.get_bytes()).digest()\n",
-    "R1_digest = hashlib.sha256(R1.get_bytes()).digest()\n",
-    "R2_digest = hashlib.sha256(R2.get_bytes()).digest()\n",
+    "# Round 1: Generate nonce point commitments and exchange them\n",
+    "# Method: use sha256() on the nonce point. sha256() takes a bytes object, so extract the bytes from the nonce point.\n",
+    "R1_digest =  # TODO: implement\n",
+    "R2_digest =  # TODO: implement\n",
+    "R3_digest =  # TODO: implement\n",
+    "\n",
+    "# Round 2: Exchange the nonce points. Each participant verifies that the nonce point commitment matches the nonce point.\n",
+    "assert R1_digest.hex() == \"05e838644f61ea13b069ca74fadd6ffa5059301663f9037704089bf3979d9e76\"\n",
+    "assert R2_digest.hex() == \"8a1467c77748b36112aeb9a73dddcc882455584324ee0769a91c530b75a22f2d\"\n",
+    "assert R3_digest.hex() == \"995a34bf709e1b620e73613d8f3462b54f47048414a6e3cbce8bec2e7fb1a24d\"\n",
     "\n",
     "# Aggregate nonces\n",
-    "# Method: use aggregate_schnorr_nonces()\n",
-    "# Negate individual nonces if necessary\n",
-    "R_agg, negated = # TODO: implement\n",
+    "# Tip: Add the individual nonce points together. If the aggregated nonce is not a quadratic residue\n",
+    "# then negate the aggregate nonce and individual nonce scalars and nonce points.\n",
+    "R_agg =  # TODO: implement\n",
+    "if  # TODO: implement\n",
+    "\n",
+    "\n",
+    "\n",
+    "\n",
+    "\n",
+    "print(\"Individual nonce scalars:\\n\\t{}, \\n\\t{}, \\n\\t{}.\\n\".format(k1, k2, k3))\n",
+    "print(\"Aggregate nonce point: {}\\n\".format(R_agg))\n",
+    "\n",
+    "# Test your solution against the aggregate_schnorr_nonces() helper function.\n",
+    "# aggregate_schnorr_nonces() aggregates the nonces and returns whether the individual nonces need to be negated.\n",
+    "test_k1, test_k2, test_k3 = k1, k2, k3\n",
+    "test_R_agg, negated = aggregate_schnorr_nonces([R1, R2, R3])\n",
     "if negated:\n",
-    "    k0.negate()\n",
-    "    k1.negate()\n",
-    "    k2.negate()\n",
+    "    test_k1 = test_k1.negate()\n",
```

Comment:
I think `negate()` doesn't return anything?

Reply:
You're correct. This is an error. Thanks!